src/expiring_lru.rs

```diff
@@ -0,0 +1,264 @@
+use tokio::time::{interval, Instant};
+use tracing::debug;
+
+pub struct Entry<V> {
+    inner: V,
+    last_used: Instant,
+}
+
+enum EntryRefInner<'a, K, V> {
+    RefMut(RefMut<'a, K, Entry<V>>),
+    RefMulti(RefMulti<'a, K, Entry<V>>),
+}
+
+pub struct EntryRef<'a, K, V>(EntryRefInner<'a, K, V>);
+
+impl<'a, K, V> EntryRef<'a, K, V>
+where
+    K: Eq + Hash,
+{
+    pub fn key(&self) -> &K {
+        match &self.0 {
+            EntryRefInner::RefMut(inner) => inner.key(),
+            EntryRefInner::RefMulti(inner) => inner.key(),
+        }
+    }
+
+    pub fn value(&self) -> &V {
+        match &self.0 {
+            EntryRefInner::RefMut(inner) => &inner.value().inner,
+            EntryRefInner::RefMulti(inner) => &inner.value().inner,
+        }
+    }
+
+    #[allow(unused)]
+    pub fn last_used(&self) -> &Instant {
+        match &self.0 {
+            EntryRefInner::RefMut(inner) => &inner.value().last_used,
+            EntryRefInner::RefMulti(inner) => &inner.value().last_used,
+        }
+    }
+}
+
+impl<'a, K, V> AsRef<V> for EntryRef<'a, K, V>
+where
+    K: Eq + Hash,
+{
+    fn as_ref(&self) -> &V {
+        match &self.0 {
+            EntryRefInner::RefMut(inner) => &inner.value().inner,
+            EntryRefInner::RefMulti(inner) => &inner.value().inner,
+        }
+    }
+}
+
+impl<'a, K, V> Deref for EntryRef<'a, K, V>
+where
+    K: Eq + Hash,
+{
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        match &self.0 {
+            EntryRefInner::RefMut(inner) => &inner.value().inner,
+            EntryRefInner::RefMulti(inner) => &inner.value().inner,
+        }
+    }
+}
+
+async fn expire_entries<K, V>(map: ExpiringLru<K, V>, reap_interval: Duration, expiration: Duration)
+where
+    K: Eq + Hash,
+{
+    let mut interval = interval(reap_interval);
+
+    loop {
+        interval.tick().await;
+
+        let right_now = Instant::now();
+        map.inner
+            .retain(|_, entry| entry.last_used + expiration > right_now);
+
+        debug!("Done reaping timed out HTTP ratelimiters");
+    }
+}
+
+pub struct ExpiringLru<K, V> {
+    inner: Arc<DashMap<K, Entry<V>>>,
+    max_size: Option<usize>,
+}
+
+impl<K, V> Clone for ExpiringLru<K, V> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            max_size: self.max_size,
+        }
+    }
+}
+
+impl<K, V> ExpiringLru<K, V>
+where
+    K: Eq + Hash + Clone + Send + Sync + 'static,
+    V: Send + Sync + 'static,
+{
+    fn new(reap_interval: Duration, expiration: Duration, max_size: Option<usize>) -> Self {
+        let inner = Arc::new(DashMap::new());
+
+        let this = Self { inner, max_size };
+
+        tokio::spawn(expire_entries(this.clone(), reap_interval, expiration));
+
+        this
+    }
+
+    pub fn insert(&self, key: K, value: V) {
+        match self.max_size {
+            Some(max_size) if max_size == 0 => return,
```

Comment:
Should this even be legal? What's the use case? If there's none, make it an illegal value inside of the builder.

Reply:
Users can set the size to 0 via an environment variable to disable the LRU cache entirely. Rather than wrapping it in an option in that case, setting the max size to 0 is the easier approach.

Reply:
Sure, I guess it disables everything but the default ratelimiter token